Handle internal links

preetmishra opened the pull request:
The PR adds support for handling internal links - stream, topic and near narrow.

#### Commits
The commits introduce `MessageLinkButton` in `MsgInfoView` and then incrementally add support for different narrows.

I would greatly appreciate any feedback on the changes, the commit chronology and the structure in general.

Partially fixes #352 and #764.

preetmishra commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -25,7 +25,6 @@ def __init__(self, controller: Any, caption: str,
             prefix = prefix_character[1]
         else:
             prefix = prefix_character
-        assert len(prefix) in (0, 1)
```
What's the reasoning behind this particular assert? The buttons don't render properly if the prefix character is >1.

I need numeric (and incremental) `prefix_character` for in-line message links which could be >1 (e.g. 10).

neiljp replied:
This defines how the current button works, ie. it is not guaranteed to work outside those bounds. If you wish it to be able to handle more, then we likely want to refactor the class to work with longer text.

That said, you're not using the extra feature which the prefix allows, which from memory is just a specific style. If you don't need that then you can skip the prefix and just put it all in the text?

preetmishra replied:
Thanks for the pointer. Given the issues with `caption`, `prefix_character` and that the existing methods weren't needed, I ended up implementing a new `urwid.Button` subclass to better tailor my needs for the message links.

preetmishra commented:
I was meaning to change the `esc` focus for successful internal link narrows but I see the current behaviour aligns with `esc` key behaviour that we already have elsewhere (e.g. web app and zulip-term itself), that is, `esc` transitions to 'All messages' narrow no matter in which narrow you are in at that moment. Thoughts?

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1023,7 +1026,28 @@ def __init__(self, controller: Any, msg: Message, title: str) -> None:
 
+        # Render the category using the existing table methods if links exist.
+        if message_links:
+            msg_info.append(('Message Links', []))
```
I am currently showing `message_links` at the bottom since it is straightforward to implement and we were planning to move reactions away from `MsgInfoView` in #707. Though, I am not sure when that would be integrated. Thoughts?

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1023,7 +1026,28 @@ def __init__(self, controller: Any, msg: Message, title: str) -> None:
+                    urwid.AttrMap(urwid.Text(caption), display_attr)
+                )
+
+            widgets.extend(message_link_widgets)
```
I am currently showing `message_links` at the bottom since it is straightforward to implement and we were planning to move reactions away from `MsgInfoView` in #707. Though, I am not sure when that would be integrated. Thoughts?

preetmishra replied:
Resolved in PM. (We're moving ahead with `message_links` before 'Reactions'.)

sumanthvrao commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1002,14 +1003,19 @@ def __init__(self, controller: Any, color: str,
 class MsgInfoView(PopUpView):
-    def __init__(self, controller: Any, msg: Message, title: str) -> None:
+    def __init__(self, controller: Any, msg: Message, title: str,
+                 message_links: 'OrderedDict[str, Tuple[str, int, bool]]',
```
If this can be None, then `Optional[MESSAGE_LINKS]`?

preetmishra replied:
This can be an empty `OrderedDict` but I believe mypy covers that currently?

preetmishra commented on zulipterminal/core.py:
```diff
@@ -134,9 +135,12 @@ def show_help(self) -> None:
 
-    def show_msg_info(self, msg: Message) -> None:
+    def show_msg_info(self, msg: Message,
+                      message_links: 'OrderedDict[str, Tuple[str, int, bool]]',
```
Apart from the conversation that was had in PM, `OrderedDict` typing doesn't work as a type alias or var (see [mypy bug](https://github.com/python/mypy/issues/6904)).

preetmishra commented:
@sumanthvrao Thanks for the review. As discussed in the PM, the first three commits are merged! :tada: 

I have pushed a rebased version for further review.

preetmishra commented:
Updated to resolve conflicts.

preetmishra commented:
Updated with tests.

neiljp commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -288,11 +288,36 @@ def handle_link(self, *_: Any) -> None:
+    def handle_narrow(self) -> None:
+        """
+        Handles stream narrow.
+        """
```
The commit title (description) for this could be improved - imagine reading just that line after coming back to it - what does this actually do? (without reading the code)

neiljp commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -288,11 +288,36 @@ def handle_link(self, *_: Any) -> None:
+        # TODO: Handle topic narrow.
+        link_split = self.link.split('/')
+
+        self.stream_id = int(link_split[5].split('-')[0])
```
Parsing the URL might be more clear here?

neiljp commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -288,11 +288,36 @@ def handle_link(self, *_: Any) -> None:
+            self.view.set_footer_text(text, duration=3)
+            return
+
+        self.stream_name = self.model.stream_dict[self.stream_id]['name']
```
Is this and `self.stream_id` for `narrow_to_stream`?

If so, it may be worth adding a comment with the assert statements.

neiljp commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -288,11 +288,36 @@ def handle_link(self, *_: Any) -> None:
+            self.controller.narrow_to_stream(self)
+
+        # Exit popup if the narrow was successful.
+        if hasattr(self.controller.loop.widget, '_Overlay__super'):
```
This seems rather cryptic - is there a better expression for this? What narrow failure are we expecting? Should the narrow methods return a value instead?

preetmishra replied:
Fixed with a simple expression. I was only expecting failure from the code that I previously had here (early return statements), but I have improved the structure now.

Re narrow methods, I'd be inclined towards handle failures, if any, in the narrow methods itself than returning a value as a separate amendment.

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -2755,3 +2755,45 @@ def test_update_widget(self, mocker, caption, expected_cursor_position,
+        (SERVER_URL + '/#narrow/stream/1-Stream-1', True),
+        (SERVER_URL + '/user_uploads/some/path/image.png', False),
+        ('https://foo.com', False),
+    ])
```
ids are always good :)

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -2755,3 +2755,45 @@ def test_update_widget(self, mocker, caption, expected_cursor_position,
+    @pytest.mark.parametrize('link, handle_narrow_called', [
+        (SERVER_URL + '/#narrow/stream/1-Stream-1', True),
+        (SERVER_URL + '/user_uploads/some/path/image.png', False),
+        ('https://foo.com', False),
```
`(SERVER_URL + '/#narrow/stream/asdasdgasdgasdg', ?),` could be in a message typed in manually, I think? What should happen?

neiljp commented on zulipterminal/helper.py:
```diff
@@ -618,3 +619,9 @@ def display_error_if_present(response: Dict[str, Any], controller: Any
         controller.view.set_footer_text(response['msg'], 3)
+
+
+def hash_util_decode(string: str) -> str:
+    # Acknowledge custom string replacements in zulip/zulip's
+    # zerver/lib/url_encoding.py before unquote.
+    return unquote(string.replace('.', '%').replace('%2E', '.'))
```
I seem to recall a discussion about wanting decoding of server URLs previously on czo - might be worth a check, maybe it's something that could be added to the python API as a utility?

preetmishra replied:
(Being discussed in [#integrations > Decode server URLs](https://chat.zulip.org/#narrow/stream/127-integrations/topic/Decode.20server.20URLs).)

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -2772,12 +2772,15 @@ def test_handle_link(self, mocker, link, handle_narrow_called):
+                              'stream_name', 'near'], [
+        (SERVER_URL + '/#narrow/stream/1-Stream-1', True, 1, 'Stream 1', None),
+        (SERVER_URL + '/#narrow/stream/4-Stream-4', False, 4, None, None),
+        (SERVER_URL + '/#narrow/stream/1-Stream-1/near/2', True, 1, 'Stream 1',
```
Could you describe what near vs not-near means in the commit, and maybe in the ids?

preetmishra commented:
@neiljp Thanks for the review and the suggestions. In the latest update, I have incorporated the in-line suggestions, improved the error handling, improved the tests with `ids` and restructured the code radically.

The PR is comparatively bigger now. Though, hopefully, the changes should be easy to comprehend with docstrings, comments, commit descriptions and tests.

neiljp commented:
@preetmishra I'm not sure if you intended to only push one commit, but seeing it alone suggests we can minimize it:
* we can simply leave `handle_link` as an empty placeholder for now? (or urlparse it more cleanly)
* we can skip storing the model and view for this first version, and add them as necessary
* the tests could go into a new `tests/ui_tools/test_buttons.py` given it's a new class (we could have started doing this sooner)

We also don't use the controller, but we know that's going to be used :)

preetmishra commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -288,11 +289,92 @@ def handle_link(self, *_: Any) -> None:
+        if not self.stream_name:
+            # FIXME: Narrow to the concerened stream in a 'preview' mode or
+            # report whether the stream id is invalid itself at least.
+            return 'You are not subscribed to the stream'
```
Thanks. I have extracted `Model.is_user_subscribed_to_stream` as method in the latest update.

Yes, we need `stream_name`.

preetmishra commented:
@neiljp Thanks for the review and extended guidelines. :+1: 

I have restructured the PR entirely to handle the links in stages that you suggested, extracted `Model.is_user_subscribed_to_stream`, improved the error handling, added support for deprecated URLs and incorporated other suggestions that left you regarding the first commit.

However, there's a caveat! The stream names that have `-`  don't get decoded appropriately. I have left a query in [development help > Encoded stream names](https://chat.zulip.org/#narrow/stream/49-development-help/topic/Encoded.20stream.20names).

preetmishra commented:
Updated with tests and rebase onto upstream/master.

neiljp commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -347,6 +359,36 @@ def _parse_narrow_link(cls, link: str) -> ParsedNarrowLink:
+            stream_id, stream_name = cls._decode_stream_data(fragments[2])
+            message_id = cls._decode_message_id(fragments[4])
+            parsed_link = {'narrow': 'stream:near'}
+            if stream_name:
+                parsed_link['stream_name'] = stream_name
+            if stream_id:
+                parsed_link['stream_id'] = stream_id
+            if message_id:
+                parsed_link['message_id'] = message_id
```
Rather than conditionally adding each of these, perhaps add them unconditionally (generate the full dict each time) and handle the validation slightly differently?

neiljp commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -418,13 +467,30 @@ def _switch_narrow_to(self, parsed_link: ParsedNarrowLink) -> None:
+                cast(Optional[int], parsed_link.get('stream_id')),
+                cast(Optional[str], parsed_link.get('stream_name')),
+            )
+            self.message = dict()
+            self.message['id'] = parsed_link['message_id']
```
We can combine these lines? (and a few similar others)

preetmishra commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -272,4 +280,136 @@ def handle_link(self, *_: Any) -> None:
+
+        narrow = cast(str, parsed_link['narrow'])
+
+        if 'stream' in narrow:
+            stream_name = parsed_link.get('stream_name')
+            stream_id = parsed_link.get('stream_id')
+
+            # Validate stream ID.
+            if stream_id and not self.model.is_user_subscribed_to_stream(
+               stream_id):
+                # TODO: Narrow to the concerened stream in a 'preview' mode or
+                # report whether the stream id is invalid instead.
+                return 'The stream seems to be either unknown or unsubscribed'
+
+            # Validate stream name.
+            if stream_name and not self.model.is_valid_stream(stream_name):
+                return 'Invalid stream name'
+
```
@neiljp Are you suggesting it to be moved to `_get_steam_data()`? I think it fits better here for two reasons:
* The overall parse -> validate -> narrow flow seems consistent.
* It is clearer to validate topic name later.

preetmishra replied:
Nevermind, I see how what you're suggesting is better. Thanks!

preetmishra replied:
Ah, I see how what you're suggesting fits better. Thanks!

preetmishra commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -249,10 +253,23 @@ def __init__(self, user_id: int, email: str) -> None:
+
+ParsedNarrowLink = TypedDict('ParsedNarrowLink', {
+    'narrow': str,
+    'stream': StreamData,
```
The following was confusing `mypy`:
```python3
ParsedNarrowLink = TypedDict('ParsedNarrowLink', {
    'narrow': str,
    'stream_id': Optional[int],
    'stream_name': Optional[str],
...
stream_data = cls._decode_stream_data(fragments[2])
parsed_link = dict(narrow='stream', **stream_data)  # type: ParsedNarrowLink
```

So, to assist `mypy` and make things simpler I introduced a `TypedDict` for the stream data and nested it instead.

preetmishra commented:
@neiljp Thanks for the suggestions. They make the code clearer and concise.

In the latest update, I have incorporated the suggestions that you left in-line and amended the tests accordingly.

neiljp commented:
@preetmishra Thanks for working on this - this is another feature that will make ZT easier to use for jumping between streams/topics :+1: This took a while to review due to the size, and there are a few refactors we can do to improve this, but I'm going to merge this now so we can get it out for wider view :tada:
